File: packages/depz-sensor-sdk-csharp/src/Depz.Sensor/Vl53l8/Cnh.cs
using System.Buffers.Binary;

namespace Depz.Sensor.Vl53l8;

/// <summary>
/// The <see cref="Vl53l8Cnh"/> input: the two aggregate/histogram dimensions of
/// the on-device CNH buffer (<c>VL53LMZ_Motion_Configuration</c>). These fix the
/// block's internal offsets, so the decode needs them alongside the raw bytes.
/// </summary>
/// <param name="NbOfAggregates">Number of CNH aggregates (device zones merged into
/// one histogram); 1..64.</param>
/// <param name="FeatureLength">CNH bins per aggregate (histogram length).</param>
public sealed record Vl53l8CnhConfig(int NbOfAggregates, int FeatureLength);

/// <summary>
/// One decoded CNH aggregate histogram. The float value of bin <c>i</c> is
/// <c>HistRaw[i] / 2^HistScaler[i]</c> (a per-bin block-floating-point mantissa +
/// shift). Both arrays are <see cref="Vl53l8CnhConfig.FeatureLength"/> long.
/// </summary>
/// <param name="HistRaw">Per-bin integer mantissa (FEAT_INT, signed int32).</param>
/// <param name="HistScaler">Per-bin base-2 scaler (FEAT_FRAC, signed int8).</param>
public sealed record Vl53l8CnhAggregate(int[] HistRaw, sbyte[] HistScaler);

/// <summary>
/// A decoded CNH data block: the reference-residual word plus one histogram per
/// aggregate (in aggregate-id order).
/// </summary>
/// <param name="RefResidualWord">Raw ref-residual word (the on-device value is
/// <c>RefResidualWord / 2048.0</c>); present because the config enables
/// MI_SFE_STORE_REF_RESIDUAL.</param>
/// <param name="Aggregates">One entry per aggregate, length
/// <see cref="Vl53l8CnhConfig.NbOfAggregates"/>.</param>
public sealed record Vl53l8CnhResult(uint RefResidualWord, IReadOnlyList<Vl53l8CnhAggregate> Aggregates);

/// <summary>
/// VL53L8<b>CH</b>-specific CNH (Compact Network Histogram) decode.
///
/// CNH is what the VL53L8CH firmware adds on top of the CX base: a per-aggregate
/// distance histogram captured in poll-mode alongside the normal ranging frame.
/// The shared results-frame decode (<see cref="Vl53l8FrameDecoder"/>) and the
/// advanced DCI codecs (<see cref="Vl53l8Advanced"/> / <see cref="MotionConfig"/>)
/// already serve both CX and CH; this is the CH-only histogram parse.
///
/// A 1:1 port of the decode path of <c>vl53lmz_plugin_cnh.c</c>
/// (<c>vl53lmz_cnh_get_block_addresses</c> / <c>_cnh_get_mem_block_addresses</c>,
/// VL53LMZ ULD 2.0.16) for the fixed <c>cnh_cfg</c> used by the DEPZ firmware
/// (DISABLE_PING_PONG | DISABLE_VARIANCE | AMBIENT | XTALK | ZERO_INVALID |
/// STORE_REF_RESIDUAL). Verified byte-exact against the live-hardware golden
/// vector <c>contracts/vectors/vl53l8_cnh.json</c>. The offset arithmetic mirrors
/// the C plugin verbatim; do not "simplify" it.
/// </summary>
public static class Vl53l8Cnh
{
    /// <summary>The variant this histogram block belongs to (always CH).</summary>
    public const Vl53l8Variant Variant = Vl53l8Variant.Ch;

    // ---- persistent-data header layout (plugin_cnh.c) ----
    private const int CnhPerHeaderWords = 5;         // CNH_PER_HEADER_BYTES / 4
    private const int CnhPerBufferHeaderWords = 2;   // CNH_PER_BUFFER_HEADER_BYTES / 4
    private const int CnhPerHeaderBufferInfoIdx = 1;
    private const int CnhPerHeaderFlagsIdx = 3;
    private const uint BufferInfoWordsMask = 0xFFFF;
    private const int MiStatePing = 0;

    /// <summary>
    /// Decode a captured CNH data block (<paramref name="block"/>, byte-swapped
    /// exactly like the standard ranging blocks) into per-aggregate histograms.
    ///
    /// Layout (little-endian words) — a faithful port of
    /// <c>_cnh_get_mem_block_addresses</c> for the fixed cnh_cfg (ping-pong +
    /// variance disabled):
    /// <list type="number">
    ///   <item><c>ref_residual_word = word[2]</c> (byte offset 8).</item>
    ///   <item>Per-buffer base = <c>CNH_PER_HEADER_WORDS (5)</c> + the ping-pong
    ///   buffer size (from the header buffer-info word); with ping-pong disabled
    ///   the device reports one buffer and this resolves to the single buffer.</item>
    ///   <item>Data starts after the 2-word buffer header, then:
    ///   FEAT_INT <c>int32[nb_agg*feat]</c>, FEAT_FRAC <c>sbyte[nb_agg*feat]</c>
    ///   (4-byte padded), AMBIENT_INT <c>int32[nb_agg]</c>, AMBIENT_FRAC
    ///   <c>sbyte[nb_agg]</c>. Aggregate <c>a</c>'s slice starts at
    ///   <c>a*feat</c>.</item>
    /// </list>
    /// </summary>
    /// <param name="config">Aggregate/bin dimensions of the CNH buffer.</param>
    /// <param name="block">Raw CNH data block bytes.</param>
    /// <returns>The ref-residual word and per-aggregate histograms.</returns>
    /// <exception cref="ArgumentOutOfRangeException">If the config dimensions are
    /// non-positive.</exception>
    /// <exception cref="ArgumentException">If <paramref name="block"/> is too short
    /// for the described layout.</exception>
    public static Vl53l8CnhResult DecodeHistogram(Vl53l8CnhConfig config, ReadOnlySpan<byte> block)
    {
        ArgumentNullException.ThrowIfNull(config);
        int nbAgg = config.NbOfAggregates;
        int feat = config.FeatureLength;
        if (nbAgg <= 0)
            throw new ArgumentOutOfRangeException(nameof(config), nbAgg, "nb_of_aggregates must be positive");
        if (feat <= 0)
            throw new ArgumentOutOfRangeException(nameof(config), feat, "feature_length must be positive");

        // A signed int32 view of the block's leading header words (LE).
        if (block.Length < CnhPerHeaderWords * 4)
            throw new ArgumentException(
                $"CNH block too short for header: {block.Length} B < {CnhPerHeaderWords * 4} B", nameof(block));

        int state = Word(block, 0);
        uint info = (uint)Word(block, CnhPerHeaderBufferInfoIdx);
        uint headerFlags = (uint)Word(block, CnhPerHeaderFlagsIdx);
        uint ppSize = info & BufferInfoWordsMask;

        // vl53lmz_cnh_get_ref_residual: raw word[2] (device value is this / 2048).
        uint refResidualWord = (uint)Word(block, 2);

        // Select ping or pong buffer exactly as the C code does. With ping-pong
        // disabled the device reports a single buffer and this resolves to ping.
        int localPp = 1;
        if ((headerFlags & 0x10) == 0x10)
            localPp = 1;
        if (state == MiStatePing)
            localPp = 1 - localPp;

        long baseWords = CnhPerHeaderWords;
        if (localPp == 1)
            baseWords += ppSize;
        // Buffer header is 2 words (state, nb_accumulated); data starts after it.
        long dataByte = (baseWords + CnhPerBufferHeaderWords) * 4;

        long aggXFeat = (long)nbAgg * feat;
        long featIntBase = dataByte;                       // FEAT_INT: int32 per (agg, feat)
        long featFracBase = featIntBase + aggXFeat * 4;    // FEAT_FRAC: sbyte per (agg, feat)

        // Bounds: the deepest read is FEAT_FRAC of the last aggregate's last bin.
        long need = featFracBase + aggXFeat;
        if (block.Length < need)
            throw new ArgumentException(
                $"CNH block too short for {nbAgg} aggregates x {feat} bins: {block.Length} B < {need} B",
                nameof(block));

        var aggregates = new Vl53l8CnhAggregate[nbAgg];
        for (int aggId = 0; aggId < nbAgg; aggId++)
        {
            long aggOff = (long)aggId * feat;
            var histRaw = new int[feat];
            var histScaler = new sbyte[feat];
            int intAt = checked((int)(featIntBase + aggOff * 4));
            int fracAt = checked((int)(featFracBase + aggOff));
            for (int i = 0; i < feat; i++)
            {
                histRaw[i] = BinaryPrimitives.ReadInt32LittleEndian(block.Slice(intAt + i * 4, 4));
                histScaler[i] = (sbyte)block[fracAt + i];
            }
            aggregates[aggId] = new Vl53l8CnhAggregate(histRaw, histScaler);
        }

        return new Vl53l8CnhResult(refResidualWord, aggregates);
    }

    /// <summary>Read the little-endian signed int32 word at <paramref name="wordIndex"/>.</summary>
    private static int Word(ReadOnlySpan<byte> block, int wordIndex) =>
        BinaryPrimitives.ReadInt32LittleEndian(block.Slice(wordIndex * 4, 4));
}

File: packages/depz-sensor-sdk-rust/src/vl53l8/cnh.rs
//! VL53L8CH CNH (Compact Network Histogram) decode — **CH-only** (contract 04).
//!
//! CNH is the reason to run the VL53L8CH firmware instead of plain VL53L8CX: the
//! sensor returns a per-aggregate distance histogram on top of the normal
//! ranging frame. The raw CNH output block is surfaced by the frame decoder
//! ([`crate::vl53l8::decode::Vl53l8Results::cnh_raw`]); this module unpacks it
//! into per-aggregate integer histograms.
//!
//! This is a 1:1 port of the decode half of the Python
//! `depz_sensor_sdk.vl53l8.cnh` (`decode` / `_decode_aggregate`), which in turn
//! ports `vl53lmz_cnh_get_block_addresses` / `_cnh_get_mem_block_addresses` from
//! the VL53LMZ ULD 2.0.16 CNH plugin for the fixed `cnh_cfg`
//! (ping-pong + variance disabled) used by DEPZ firmware. The offset arithmetic
//! is reproduced exactly; do not "simplify" it.
//!
//! Layout of the byte-swapped CNH block (little-endian 32-bit words):
//! - word[2] (byte offset 8) = `ref_residual_word` (u32).
//! - a 5-word persistent header, then one (ping-pong disabled) buffer whose size
//!   in words is read from the header's buffer-info word; each buffer starts
//!   with a 2-word buffer header, followed by:
//!     - `FEAT_INT`   : `i32[nb_agg * feat]`
//!     - `FEAT_FRAC`  : `i8 [nb_agg * feat]` (padded up to a word boundary)
//!     - `AMBIENT_INT`: `i32[nb_agg]`
//!     - `AMBIENT_FRAC`: `i8[nb_agg]`
//!   The real histogram value is `hist_raw / 2**hist_scaler`.

// ---- persistent-data header layout (plugin_cnh.c) ----
const CNH_PER_HEADER_WORDS: usize = 5; // CNH_PER_HEADER_BYTES / 4
const CNH_PER_BUFFER_HEADER_WORDS: usize = 2; // CNH_PER_BUFFER_HEADER_BYTES / 4
const CNH_PER_HEADER_BUFFER_INFO_IDX: usize = 1;
const CNH_PER_HEADER_FLAGS_IDX: usize = 3;
const BUFFER_INFO_WORDS_MASK: u32 = 0xFFFF;
const BUFFER_INFO_FLAGS_SHIFT: u32 = 24;
const MI_STATE_PING: i32 = 0;

/// Minimal config needed to decode a captured CNH block: the aggregate count and
/// per-aggregate feature (bin) length the sensor was configured with. These must
/// match the `CnhConfig` used when programming the device (see the Python
/// `CnhConfig.nb_of_aggregates` / `feature_length`).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CnhDecodeConfig {
    /// Number of CNH aggregates (`cfg.nb_of_aggregates`).
    pub nb_of_aggregates: usize,
    /// CNH bins per aggregate (`cfg.feature_length`).
    pub feature_length: usize,
}

/// One decoded CNH aggregate. The real histogram value for bin `i` is
/// `hist_raw[i] as f64 / 2f64.powi(hist_scaler[i] as i32)`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CnhAggregate {
    /// Per-bin integer mantissa (`FEAT_INT`), length == `feature_length`.
    pub hist_raw: Vec<i32>,
    /// Per-bin power-of-two scaler (`FEAT_FRAC`), length == `feature_length`.
    pub hist_scaler: Vec<i8>,
}

/// Result of [`decode_cnh`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CnhData {
    /// Reference residual word, u32 at byte offset 8 (`words[2]`). Real value is
    /// `ref_residual_word as f64 / 2048.0` (11 fractional bits).
    pub ref_residual_word: u32,
    /// Per-aggregate histograms, length == `nb_of_aggregates`.
    pub aggregates: Vec<CnhAggregate>,
}

/// CNH decode failure.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CnhError {
    /// `raw` is too short for the header or the computed block extends past it.
    Truncated,
    /// `nb_of_aggregates` or `feature_length` is zero.
    EmptyConfig,
}

impl std::fmt::Display for CnhError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CnhError::Truncated => write!(f, "VL53L8 CNH block truncated"),
            CnhError::EmptyConfig => write!(f, "VL53L8 CNH empty config"),
        }
    }
}

impl std::error::Error for CnhError {}

/// Read a signed little-endian i32 word at byte offset `off`.
#[inline]
fn read_i32(raw: &[u8], off: usize) -> Result<i32, CnhError> {
    let end = off.checked_add(4).ok_or(CnhError::Truncated)?;
    let bytes = raw.get(off..end).ok_or(CnhError::Truncated)?;
    Ok(i32::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3]]))
}

/// Read a signed i8 at byte offset `off`.
#[inline]
fn read_i8(raw: &[u8], off: usize) -> Result<i8, CnhError> {
    raw.get(off).map(|&b| b as i8).ok_or(CnhError::Truncated)
}

/// Decode a captured CNH data block (`raw` bytes, byte-swapped exactly like the
/// standard ranging blocks — i.e. [`crate::vl53l8::decode::Vl53l8Results::cnh_raw`])
/// into per-aggregate integer histograms plus the reference-residual word.
///
/// Faithful port of the Python `cnh.decode` / `_decode_aggregate` for the fixed
/// DEPZ `cnh_cfg` (ping-pong + variance disabled). With ping-pong disabled the
/// device reports a single buffer and the ping/pong selection resolves to the
/// sole buffer.
pub fn decode_cnh(cfg: &CnhDecodeConfig, raw: &[u8]) -> Result<CnhData, CnhError> {
    let nb_agg = cfg.nb_of_aggregates;
    let feat = cfg.feature_length;
    if nb_agg == 0 || feat == 0 {
        return Err(CnhError::EmptyConfig);
    }

    // ref_residual: u32 at word[2] (byte offset 8) — vl53lmz_cnh_get_ref_residual.
    let ref_residual_word = read_i32(raw, 2 * 4)? as u32;

    let mut aggregates = Vec::with_capacity(nb_agg);
    for agg_id in 0..nb_agg {
        aggregates.push(decode_aggregate(raw, nb_agg, feat, agg_id)?);
    }

    Ok(CnhData {
        ref_residual_word,
        aggregates,
    })
}

fn decode_aggregate(
    raw: &[u8],
    nb_agg: usize,
    feat: usize,
    agg_id: usize,
) -> Result<CnhAggregate, CnhError> {
    let agg_x_feat = nb_agg * feat;
    let agg_off = agg_id * feat;

    let state = read_i32(raw, 0)?;
    let info = read_i32(raw, CNH_PER_HEADER_BUFFER_INFO_IDX * 4)? as u32;
    let pp_size = (info & BUFFER_INFO_WORDS_MASK) as usize;
    let _buffer_flags = (info >> BUFFER_INFO_FLAGS_SHIFT) & 0xFF;

    // Select ping or pong buffer exactly as the C code does. With ping-pong
    // disabled the device reports a single buffer and this resolves to ping.
    let flags = read_i32(raw, CNH_PER_HEADER_FLAGS_IDX * 4)?;
    let mut local_pp: usize = 1;
    if (flags & 0x10) == 0x10 {
        local_pp = 1;
    }
    if state == MI_STATE_PING {
        local_pp = 1 - local_pp;
    }

    let mut base = CNH_PER_HEADER_WORDS;
    if local_pp == 1 {
        base += pp_size;
    }
    // buffer header is 2 words (state, nb_accumulated); data starts after it.
    let mut blk = (base + CNH_PER_BUFFER_HEADER_WORDS) * 4; // byte offset of p[2]

    // FEAT_INT: int32 per (agg, feat).
    let feat_int_off = blk + agg_off * 4;
    let mut hist_raw = Vec::with_capacity(feat);
    for i in 0..feat {
        hist_raw.push(read_i32(raw, feat_int_off + i * 4)?);
    }
    blk += agg_x_feat * 4;

    // FEAT_FRAC: int8 scaler per (agg, feat).
    let feat_frac_off = blk + agg_off;
    let mut hist_scaler = Vec::with_capacity(feat);
    for i in 0..feat {
        hist_scaler.push(read_i8(raw, feat_frac_off + i)?);
    }
    // (AMBIENT_INT / AMBIENT_FRAC follow after ((3 + agg_x_feat) / 4) * 4 bytes;
    // not decoded here — the DEPZ CNH surface returns histograms only.)

    Ok(CnhAggregate {
        hist_raw,
        hist_scaler,
    })
}
